db/nft.go

```diff
@@ -381,135 +381,144 @@ func GetNftEvents(conn *pgxpool.Conn, q QueryEventsRequest, p PageRequest) (Quer
+					AND ($5::text[] IS NULL OR cardinality($5::text[]) = 0 OR i.address = ANY($5))
+					AND ($6 = 0 OR (e.timestamp IS NOT NULL AND e.timestamp > to_timestamp($6)))
+					AND ($7 = 0 OR (e.timestamp IS NOT NULL AND e.timestamp < to_timestamp($7)))
+					AND ($8::text[] IS NULL OR cardinality($8::text[]) = 0 OR e.action = ANY($8))
+					AND ($9 = false OR e.receiver != e.iscn_owner_at_the_time)
+					AND (%[2]s)
+				GROUP BY e.class_id, c.created_at, i.address
+			) AS sum_query
+			ORDER BY class_rank, income DESC
+		) AS rank_query
+		WHERE class_rank > $1 AND class_rank <= $2
+	`, orderBy, ownershipCondition)
 
 	ctx, cancel := GetTimeoutContext()
 	defer cancel()
 
+	rankStart := int(p.Key)
+	rankEnd := rankStart + p.Limit
 	rows, err := conn.Query(
 		ctx, sql,
-		p.Limit, p.After(), p.Before(), q.ClassId, q.NftId,
-		ownerVariations, stakeholderVariations, q.After, q.Before, q.ActionType,
+		rankStart, rankEnd, q.ClassId, ownerVariations, beneficiaryVariations,
+		q.After, q.Before, q.ActionType, q.ExcludeSelfPurchase,
 	)
 	if err != nil {
 		logger.L.Errorw("Failed to query nft incomes", "error", err)
 		return QueryIncomesResponse{}, fmt.Errorf("query nft incomes error: %w", err)
 	}
 
 	res := QueryIncomesResponse{
-		Incomes: make([]NftIncomeResponse, 0),
+		ClassIncomes: make([]NftClassIncomeResponse, 0),
 	}
+
+	var rank int
+	rankMap := make(map[int]NftClassIncomeResponse)
 	for rows.Next() {
-		var r NftIncomeResponse
-		if err = rows.Scan(&r.Address, &r.Amount); err != nil {
+		var ci NftClassIncomeResponse
+		var i NftIncomeResponse
+
+		if err = rows.Scan(&ci.ClassId, &ci.CreatedAt, &i.Address, &i.Amount, &ci.TotalAmount, &rank); err != nil {
 			logger.L.Errorw("failed to scan nft incomes", "error", err, "q", q)
 			return QueryIncomesResponse{}, fmt.Errorf("query nft incomes data failed: %w", err)
 		}
-		res.Incomes = append(res.Incomes, r)
-		res.TotalAmount += r.Amount
-	}
-	res.Pagination.Count = len(res.Incomes)
-	return res, nil
-}
 
-func GetNftIncomeDetails(conn *pgxpool.Conn, q QueryIncomeDetailsRequest, p PageRequest) (QueryIncomeDetailsResponse, error) {
-	ownerVariations := utils.ConvertAddressPrefixes(q.Owner, AddressPrefixes)
-	stakeholderVariations := utils.ConvertAddressPrefixes(q.Address, AddressPrefixes)
-	orderBy := "i.id"
-	switch q.OrderBy {
-	case "price":
-		orderBy = "e.price"
-	case "income":
-		orderBy = "i.amount"
-	case "default":
-	default:
-		orderBy = "i.id"
+		classIncome, exists := rankMap[rank]
+		if !exists {
+			classIncome = ci
+			classIncome.Incomes = make([]NftIncomeResponse, 0)
+		}
+
+		classIncome.Incomes = append(classIncome.Incomes, i)
+		rankMap[rank] = classIncome
+		res.TotalAmount += i.Amount
 	}
+	var classIds []string
+	for _, classIncome := range rankMap {
+		res.ClassIncomes = append(res.ClassIncomes, classIncome)
+		classIds = append(classIds, classIncome.ClassId)
+	}
+	for i := 0; i < len(res.ClassIncomes); i++ {
+		res.ClassIncomes[i] = rankMap[i+rankStart+1]
```

Comment:
```suggestion
		res.ClassIncomes[i] = rankMap[rankStart+i+1]
```
Personally I think this is more readable but just my preference...